ambry-api/src/main/java/com/github/ambry/protocol/RequestAPI.java

```diff
@@ -25,77 +25,88 @@ public interface RequestAPI {
   /**
    * Handle a request.
    * @param request The request to handle.
+   * @param shouldDrop indicates if request should be dropped.
    * @throws InterruptedException if request processing is interrupted.
    */
-  void handleRequests(NetworkRequest request) throws InterruptedException;
+  void handleRequests(NetworkRequest request, boolean shouldDrop) throws InterruptedException;
```

Comment:
These interfaces look weird to me for two reasons.
1. Dropping request should be done in netty io thread, it shouldn't invoke request handling requests.
2. Even if we want to do this in request handling request, can we have a new method like handleDropppedRequest?

Reply:
+1
Not sure how easy or hard to reject the requests in netty io thread since it doesn't deal with AmbryRequest before. But if it's doable, we will be able to reject these request at the earliest convenience. It's more efficient.

And with either 1 or 2, it can avoid the massive interface change.

Reply:
Would the netty thread print out any logs indicating that requests are being rejected ? or we might be in a limbo.

Reply:
I think this cannot be achieved in the netty io thread since the decision to drop a request cannot be reached during enqueuing. Requests can decay while they are in the queue.

OTOH, +1 on having a generic request drop routing rather than delaying it to specific request handlers, e.g., handlePutRequest.

Reply:
Sure. Made changes to have a separate method in RequestAPI to drop requests instead of modifying all `handlexxxRequest` methods. Also, moved metrics also to a common place so that they can be reused for both 'handled' and 'dropped' requests.

Even though it is ideal to send responses 'unqueued' requests immediately in Netty Thread, I was thinking it might be better to have same request-response flow path (i.e. requests queued by Netty thread and dequeued, processed and handed over back to netty outbound queues by request handler thread) for both 'unqueued' and 'delayed' requests. It was also easier to have response handling code (like updating metrics, etc) in one place in that way.